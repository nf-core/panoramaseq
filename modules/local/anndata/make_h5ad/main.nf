process tsv_to_h5ad {
    tag "$meta.id"
    label 'process_single'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'oras://community.wave.seqera.io/library/anndata:0.10.9--d13580e4b297da7c':
        'community.wave.seqera.io/library/anndata:0.10.9--1eab54e300e1e584' }"

    input:
    tuple val(meta), path(count_tsvs), path(coords_csv)

    output:
    tuple val(meta), path("*.h5ad"), emit: h5ad
    path "versions.yml"            , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def tsv_files = count_tsvs instanceof java.util.Collection ? count_tsvs.join(' ') : count_tsvs
    """
    tsv_to_h5ad.py \\
        ${tsv_files} \\
        --coords ${coords_csv} \\
        --output ${prefix}.h5ad \\
        ${args}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$(python3 -c 'import platform; print(platform.python_version())')
        anndata: \$(python3 -c 'import anndata as ad; print(ad.__version__)')
        pandas: \$(python3 -c 'import pandas as pd; print(pd.__version__)')
        scipy: \$(python3 -c 'import scipy; print(scipy.__version__)')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.h5ad

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$(python3 -c 'import platform; print(platform.python_version())')
        anndata: \$(python3 -c 'import anndata as ad; print(ad.__version__)')
        pandas: \$(python3 -c 'import pandas as pd; print(pd.__version__)')
        scipy: \$(python3 -c 'import scipy; print(scipy.__version__)')
    END_VERSIONS
    """
}