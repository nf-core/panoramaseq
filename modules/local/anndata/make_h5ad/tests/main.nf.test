nextflow_process {

    name "Test Process tsv_to_h5ad"
    script "../main.nf"
    process "tsv_to_h5ad"
    tag "modules"
    tag "modules_local"
    tag "tsv_to_h5ad"
    tag "make_h5ad"

    test("tsv_to_h5ad - stub") {

        options "-stub"

        when {
            params {
                outdir = "$outputDir"
                panoramaseq_testdata_base_path = "https://raw.githubusercontent.com/francops1722/test-datasets/panoramaseq/testdata/"
            }
            process {
                """
                input[0] = [
                    [ id: 'test_h5ad' ],
                    [
                        file(params.panoramaseq_testdata_base_path + 'counts/S09_counts.tsv.gz', checkIfExists: true),
                        file(params.panoramaseq_testdata_base_path + 'counts/S10_counts.tsv.gz', checkIfExists: true)
                    ],
                    file(params.panoramaseq_testdata_base_path + 'barcodes_coords.csv', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert process.out.h5ad != null },
                { assert process.out.h5ad.get(0).get(1) ==~ ".*\\.h5ad" },
                { assert process.out.versions != null },
                { assert snapshot(process.out.versions).match() }
            )
        }
    }

    test("tsv_to_h5ad - single file") {

        when {
            params {
                outdir = "$outputDir"
                panoramaseq_testdata_base_path = "https://raw.githubusercontent.com/francops1722/test-datasets/panoramaseq/testdata/"
            }
            process {
                """
                input[0] = [
                    [ id: 'single_h5ad' ],
                    file(params.panoramaseq_testdata_base_path + 'counts/S09_counts.tsv.gz', checkIfExists: true),
                    file(params.panoramaseq_testdata_base_path + 'barcodes_coords.csv', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert process.out.h5ad != null },
                { assert process.out.h5ad.get(0).get(1) ==~ ".*\\.h5ad" },
                { assert file(process.out.h5ad.get(0).get(1)).exists() },
                { assert process.out.versions != null },
                { assert snapshot(process.out.h5ad, process.out.versions).match("single_file") }
            )
        }
    }

    test("tsv_to_h5ad - multiple files") {

        when {
            params {
                outdir = "$outputDir"
                panoramaseq_testdata_base_path = "https://raw.githubusercontent.com/francops1722/test-datasets/panoramaseq/testdata/"
            }
            process {
                """
                input[0] = [
                    [ id: 'multi_h5ad' ],
                    [
                        file(params.panoramaseq_testdata_base_path + 'counts/S09_counts.tsv.gz', checkIfExists: true),
                        file(params.panoramaseq_testdata_base_path + 'counts/S10_counts.tsv.gz', checkIfExists: true)
                    ],
                    file(params.panoramaseq_testdata_base_path + 'barcodes_coords.csv', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert process.out.h5ad != null },
                { assert process.out.h5ad.get(0).get(1) ==~ ".*\\.h5ad" },
                { assert file(process.out.h5ad.get(0).get(1)).exists() },
                { assert process.out.versions != null },
                { assert snapshot(process.out.h5ad, process.out.versions).match("multiple_files") }
            )
        }
    }
}
