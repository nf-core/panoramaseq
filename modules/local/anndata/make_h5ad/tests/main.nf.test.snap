{
    "single_file": {
        "content": [
            [
                [
                    {
                        "id": "single_h5ad"
                    },
                    "single_h5ad.h5ad:md5,0213589d4fbd8b1da061b5968d6fb1fc"
                ]
            ],
            [
                "versions.yml:md5,df9ef7326600bda8c206164a377ce26a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2025-07-02T10:42:24.684048314"
    },
    "multiple_files": {
        "content": [
            [
                [
                    {
                        "id": "multi_h5ad"
                    },
                    "multi_h5ad.h5ad:md5,ea2aa3fccb7c5e96e26abe546d74127b"
                ]
            ],
            [
                "versions.yml:md5,df9ef7326600bda8c206164a377ce26a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2025-07-02T10:42:37.355745686"
    },
    "tsv_to_h5ad - stub": {
        "content": [
            [
                "versions.yml:md5,df9ef7326600bda8c206164a377ce26a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2025-07-02T10:42:17.003033403"
    }
}